### PR TITLE
Transhumanism rework: Bionic slots

### PR DESCRIPTION
**Pull request by Ramza13**
#### Summary
Features "Adjust bionic slots for first step of rework"

#### Purpose of change
First part of #28273
Begin rework of bionics to use max and safe slots and faults for overuse
With the exodites coming next version to shake up cbms, it will be the perfect time to flesh them out more.

#### Describe the solution
This needs balancing/feedback but it sets up much of the linked issues framework.

- Adds the concepts of maximum and safe slots for each body part.    Right now all max values are unchanged and all safe vals are 50% of them.
- CBM installation chances are affected by going over safe slot usage. Right now a 30% bonus is applied if no affected parts have bionics and a 30% penalty is applied if any parts are over safe limit before install.
- Adds faults.  For each body part you define in json a list of faults, which are an effect id and a number of slots.  When you use that number of slots or more on that body part you will have that effect on the body part.  Right now there are 3 faults for all body parts: open_skin which has a chance of infection, and nerve_pain and nerve_pain_2 which both add pain. These need balance/tweaking.
- Each body parts maximum slot number will be lowered by mutations affecting it.  Right now this is using existing bp tags in mutation but in the future will need new trait tags to fully capture.
- Added ability for effects to cause infections, as this was needed for open skin.
- Removes the option of not using cbm slots mode.

### CODE REVIEW

**Comment by anothersimulacrum** [comment]
> CBM installation chances are affected by going over safe slot usage. Right now a 30% bonus is applied if no affected parts have bionics and a 30% penalty is applied if any parts are over safe limit before install.

So a 60% swing?

**Comment by Ramza13** [comment]
> 
> 
> > CBM installation chances are affected by going over safe slot usage. Right now a 30% bonus is applied if no affected parts have bionics and a 30% penalty is applied if any parts are over safe limit before install.
> 
> So a 60% swing?

Yes. Its a placeholder.   I want to make it more nuanced and reasonable but wasn't really sure the best way. 
In a future PR or as further commits to this I hope to get a better system.  Do you have any thoughts?

**Comment by anothersimulacrum** [comment]
Not in particular, I just wanted to clarify that you didn't mean that the 30% bonus was removed and so it was only a 30% swing.

**Comment by actual-nh** [comment]
Very interesting!

> With the exodites coming next version to shake up cbms, it will be the perfect time to flesh them out more.

What is the current plan on this?

> * Adds the concepts of maximum and safe slots for each body part.    Right now all max values are unchanged and all safe vals are 50% of them.

I suggest the current slots need multiplying by 2-10, with slots required for most CBMs multiplied by the same amount. This will make it easier to do items with a small amount in one location but most in another (it would make sense for the nanobots to have a bit of "infrastructure" in the extremities), or smaller-scale items like the subcutaneous carbon filaments. (See #46212 for a bit more discussion on this.)

> * CBM installation chances are affected by going over safe slot usage. Right now a 30% bonus is applied if no affected parts have bionics and a 30% penalty is applied if any parts are over safe limit before install.

There are quite a few of the bionics that could be easier to install because of each other - Thermal Dissipation and Internal Climate Control, Oxygenator and Air Filtration System, etc. (For an extreme case, repair nanobots - especially if the autodoc could communicate with them! - should make most installations easier.) Even without such special cases (which may well need to wait for a later PR), a bit more gradual change than going from a 30% to nothing makes, IMO, a bit more sense.

> * Adds faults.  For each body part you define in json a list of faults, which are an effect id and a number of slots.  When you use that number of slots or more on that body part you will have that effect on the body part.  Right now there are 3 faults for all body parts: open_skin which has a chance of infection, and nerve_pain and nerve_pain_2 which both add pain. These need balance/tweaking.

Ideally, the open skin one could be confined to the cases in which something is actually going through the skin. Admittedly, a lot of this can be done via the body parts - torso would take a lot more than hands, for instance.

> * Removes the option of not using cbm slots mode.

I suggest this will be received better if using slots instead becomes the default, and a (in-repository) mod is available to turn them off.

**Comment by anothersimulacrum** [comment]
> > * Removes the option of not using cbm slots mode.
> 
> I suggest this will be received better if using slots instead becomes the default, and a (in-repository) mod is available to turn them off.

It doesn't fit our policy on in-repo mods, and it's much less painless to do it in one batch than in two.

File: data/json/body_parts.json [hunk 1]
```diff
@@ -24,7 +24,13 @@
     "base_hp": 60,
     "drench_capacity": 40,
     "smash_message": "You smash the %s with a powerful shoulder-check.",
-    "bionic_slots": 80
+    "safe_bionic_slots": 40,
+    "max_bionic_slots": 80,
+    "faults": [
+      { "effect": "open_skin", "slots_requirement": 40 },
+      { "effect": "nerve_pain", "slots_requirement": 50 },
+      { "effect": "nerve_pain_2", "slots_requirement": 60 }
```
Comment (actual-nh) [inline data/json/body_parts.json:32]:
```suggestion
      { "effect": "open_skin", "slots_requirement": 60 },
      { "effect": "nerve_pain", "slots_requirement": 40 },
      { "effect": "nerve_pain_2", "slots_requirement": 50 }
```

**Comment by actual-nh** [comment]
> It doesn't fit our policy on in-repo mods, and it's much less painless to do it in one batch than in two.

Unfortunate. Ah, well. (I don't disagree with using slots to affect the chance of successful installation - I actually had the idea myself before I saw #28273!)

**Comment by actual-nh** [comment]
Oh. One question: How will slot usage affect removal success?

File: data/json/body_parts.json (hunk 1 above)
Comment (Ramza13) [reply data/json/body_parts.json:32]: Why?

Comment (actual-nh) [reply data/json/body_parts.json:32]: It is a lot less likely that something in the torso is going to be close enough to the surface to cause skin problems (take a look at which CBMs are in the torso). OTOH, there are still a lot of nerves there...

**Comment by Ramza13** [comment]
> 
> 
> Oh. One question: How will slot usage affect removal success?

At the moment it has no effect.   I will probably start with a general bonus/penalty, and then use a similar mechanism to what the final installation bonus/penalty ends up as.

**Comment by actual-nh** [comment]
BTW, what happens re safe and max slots with things that adjust their numbers, like the Biotek mage in Manaclysm (see [traits/attunements.json#L92](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/data/mods/Magiclysm/traits/attunements.json#L92))?

**Comment by Asmitha90** [comment]
Not a big fan of safe bionic slot or overall bionic slot option. In theory you coud replace ewything the question is what kaind of material the cbm are made... Weight, encumbrenc, speed, efect debufs  seem to be a better way to do this.

**Comment by Aivean** [comment]
In my opinion, having the "safe limit" and negative effects for exceeding it makes sense. 

But, having a hard "max" limit is weird. I think human body and skin is stretchy and adaptive, it's always possible to fit N+1 thing inside. However, fitting too much should result in encumbrance increase (perhaps, exponential?) and installing too many cbms should be impractical because of that.

**Comment by Raikiri** [comment]
@Aivean if CMB's just have encumbrance/weight, they'll be too similar to equipment. I also think the point is to make it so that low-commitment CMB installations are easy no-brainers, but the further you go, the more drawbacks you get. With equipment-like encumbrance penalties, you'd get negatives for any amount installed.

**Comment by Aivean** [comment]
> Maybe start adding encumbrance penalties once you get past the limit?

That's what I meant.

My argument was about the hard "max" limit. I think that having a hard "max" cap on the CBM slots makes little sense from the reality standpoint. I believe soft "max" cap is a better solution: making it not practical for the player to cross the "max" threshold, instead of outright forbidding it.

**Comment by Asmitha90** [comment]
> @Aivean if CMB's just have encumbrance/weight, they'll be too similar to equipment. I also think the point is to make it so that low-commitment CMB installations are easy no-brainers, but the further you go, the more drawbacks you get. With equipment-like encumbrance penalties, you'd get negatives for any amount installed.

Mutualy exclusive bionic and lots of encumbrenc + weight solve ewrything since if you instal to much you cant carry anthig  or wear armor... This forces you to chose what you realy need also specialized bionic like limbs etc. is the way to go..

**Comment by anothersimulacrum** [comment]
> But, having a hard "max" limit is weird. I think human body and skin is stretchy and adaptive, it's always possible to fit N+1 thing inside. However, fitting too much should result in encumbrance increase (perhaps, exponential?) and installing too many cbms should be impractical because of that.

There are many more downsides than encumbrance to trying to fit much more in than you can in, maximum slots is a good compromise without going too far into the mud. I think it might be interesting at one point to explore those, but it'd be very safe to stick with that here and consider those later.

**Comment by Aivean** [comment]
> There are many more downsides than encumbrance to trying to fit much more in than you can in, maximum slots is a good compromise without going too far into the mud. I think it might be interesting at one point to explore those, but it'd be very safe to stick with that here and consider those later.

I'm not sure about that.

My rationale is following: if you place a large artificial object under the skin and it is not toxic and not rejected by the body, then, after relatively short healing and adaptation process, the only real downside is encumbrance. In my view the human body is adaptive enough to accommodate several times its volume (if added slowly enough).

Several examples come to mind: 
* existing subdermal implants, such as breast and butt implants
* fat deposits in obese people
* pregnancy
* synthol injections

However, I'm not a medical expert, and I could be wrong or missing something. Perhaps actual medical experts, @I-am-Erk @Venera3, could add something to the topic.

**Comment by Venera3** [comment]
Open skin as a fault name is a bit problematic, because it implies that it's confined to the skin (not that I have a better suggestion, mind). Every surface where the implant and the flesh come into contact is essentially an open wound, and with that a potential site for a foreign body reaction and bacterial colonisiation. Smaller implants like a pacemaker or those in closed bags of inert material ( the plastic surgery stuff) lead to such problems rarely, but anything with an organ "interface" is going to cause a low-level inflammation. 
Things that get hooked into/replace parts of an organ would be the main offenders ( the gut optimizer/metabolic interchange, all the nervous system upgrades, blood filter), followed by CBMs that have an external interface (most energy cbms, monomolecular blade), and the "closed bag"/inert CBMs (alloy plating, the minor stuff like a clock) would be the least problematic in that respect. That's ignoring the mass and encumberance, where the alloy plating would be more of a factor.

All in all I see the slots as representing the capacity of the limb less in the physical sense and more like "this part can take this much abuse before the remaining flesh starts falling off", so I find the systemic debuffs from going over the safe capacity to be a fitting consequence. 

Encumberance and other, more CBM-specific drawbacks belong in their definitions and would clutter up the slot system imho.

Mutations reducing the limb capacity is a nice middle ground between throwing out the CBM entirely and having no consequence to mixing both. The general "body size" mutations could also affect the slot capacity of all limbs, with the Large line raising and the Unassuming lowering it respectively.

A way for CBMs to influence the safe slot limit would also be nice, to allow for the medical things helping and some unstable/faulty CBMs  to lower it for other limbs as well ( when the real wacky stuff makes it in). But that can be a separate addition.

Sorry if that got a bit meandering, but I didn't follow the discussion closely until now.

**Comment by Asmitha90** [comment]
Do you have a coherent vision of what tech level is in cdda? 
For me this looks like this isnt clearly defined here and this why there are problems...
If we have bionic like "Repair Nanobots" then i assume we reached nano technology level. This would make most of the surgery, safe space or incompatibility problem non-existent(neural stuff and similar would still be problematic) since the surgeries cloud happen at cell level.

Second there's no defined end point/goal for this route do we go cyborg or full synthetic. I fell like here is where the talk about body should start. If a hybrid(cyborg) is where we draw the line then i agree that max slots and limitations should be in place but if we go to full synthetic this is redundant since sooner or later you will replace everything in your body(maybe head should only have this system in place).

**Comment by actual-nh** [comment]
> Open skin as a fault name is a bit problematic, because it implies that it's confined to the skin (not that I have a better suggestion, mind). Every surface where the implant and the flesh come into contact is essentially an open wound, and with that a potential site for a foreign body reaction and bacterial colonisiation. Smaller implants like a pacemaker or those in closed bags of inert material ( the plastic surgery stuff) lead to such problems rarely, but anything with an organ "interface" is going to cause a low-level inflammation.
> Things that get hooked into/replace parts of an organ would be the main offenders ( the gut optimizer/metabolic interchange, all the nervous system upgrades, blood filter),

I can see that for the gut ones (assuming they actually penetrate through the gut wall and aren't entirely inside or out), and for that matter for most designs for lung ones (could manage to build something that was entirely "outside" the body for some things, perhaps). I would suggest otherwise for some of the rest, although I can certainly see your viewpoint:

- If it's entirely inside the CNS, the regular immune system isn't a problem.
- At least by the original CBM background, they were using material for interfacing that already was able to evade immune system interference.
- By the leukocyte breeder, the tech is already able to manipulate the immune system to a significant degree.
- If need be, raise the chance of various diseases (ambient especially) to represent immunosuppression. (To counter the leukocyte breeder making this not a problem, up the risk of acquiring various allergies and autoimmune diseases (the latter being things like ROT#, IMO.)
- See below re bacterial contamination.

> followed by CBMs that have an external interface (most energy cbms, monomolecular blade)

Definitely (which would also potentially include the gut and lung ones mentioned above). Perhaps "Infection Risk"?

>, and the "closed bag"/inert CBMs (alloy plating, the minor stuff like a clock) would be the least problematic in that respect.

Installation problems would be the ones that would cause long-term side effects with those (biofilm formation, for instance) - currently, all the installation problems prevent installation, IIRC; this needs revising, IMO.

> A way for CBMs to influence the safe slot limit would also be nice, to allow for the medical things helping and some unstable/faulty CBMs to lower it for other limbs as well ( when the real wacky stuff makes it in). But that can be a separate addition.

Agreed.

**Comment by Raikiri** [comment]
> Second there's no defined end point/goal for this route do we go cyborg or full synthetic. I fell like here is where the talk about body should start. If a hybrid(cyborg) is where we draw the line then i agree that max slots and limitations should be in place but if we go to full synthetic this is redundant since sooner or later you will replace everything in your body(maybe head should only have this system in place).

honestly it'd be amazing if the game officially allowed to transcend your own humanity and actually replace your fleshy body with synthetic one over time.

**Comment by Venera3** [comment]
Re: tech level, the game started with an "everything and the kitchen sink" cyberpunk dystopia including chicken walkers and (iirc) police bots you could pay off. That morphed into the current goal of "next year, but what if secret aliens 20 years ago", making CBMs stick out more and more as off-theme. They got away with it mostly by being too cool and iconic to remove, but got bumped up to end-game content. In the nearish future they will be reworked to be of off-dimensional origin, accessible through a dedicated alien (kinda.) faction.

Brain-in a jar/ head transplant body conversions got discussed in the linked issue and are on the table as very endgame stuff with hefty disadvantages. Funnily enough, those would have little problem with the whole open skin thing, since the surfaces are relatively small in comparison to a cyborg.

The thing about gut/lung stuff is whatever the configuration, you *have* to have a contact surface to the  blood at least, and whether you do that with a direct line or through a membrane you'll run into problems. Secondary plastic/heart valve infections are a huge problem irl and can be triggered by any transient bacteria that manage to get into the blood stream.

CNS implants might be shielded from the systemic immune answer, but there is a  local immune system that may or may not be enthusiastic about a few metres of foreign cabling. Secondary infections are also a problem in those as well.

**Comment by actual-nh** [comment]
> Re: tech level, the game started with an "everything and the kitchen sink" cyberpunk dystopia including chicken walkers and (iirc) police bots you could pay off. That morphed into the current goal of "next year, but what if secret aliens 20 years ago", making CBMs stick out more and more as off-theme. They got away with it mostly by being too cool and iconic to remove, but got bumped up to end-game content. In the nearish future they will be reworked to be of off-dimensional origin, accessible through a dedicated alien (kinda.) faction.

Exodii, yes.

> Brain-in a jar/ head transplant body conversions got discussed in the linked issue and are on the table as very endgame stuff with hefty disadvantages. Funnily enough, those would have little problem with the whole open skin thing, since the surfaces are relatively small in comparison to a cyborg.

Chuckle. Yes.

> The thing about gut/lung stuff is whatever the configuration, you _have_ to have a contact surface to the blood at least, and whether you do that with a direct line or through a membrane you'll run into problems. Secondary plastic/heart valve infections are a huge problem irl and can be triggered by any transient bacteria that manage to get into the blood stream.

Yeah. I do think that may be solvable by coating it with a set of cells (3D cell printing).

> CNS implants might be shielded from the systemic immune answer, but there is a local immune system that may or may not be enthusiastic about a few metres of foreign cabling. Secondary infections are also a problem in those as well.

True, as with cochlear implants. (One interesting idea for an Exodii implant: Boosting the brain's local immune system, such as the chitosan production (something I ran across in my dissertation work) - looks to help protect against fungal and possibly insect infections by degrading their chitin!)

**Comment by Aivean** [comment]
Thank you guys for a very useful input.

After reading Venera's comment I understood why slots seemed non-sensical to me. Currently it seems that they basically represent the volume of the CBM, which means that large and mostly inert CBMs take up too much of this limit.

I think we should rename "slots" to something like "invasiveness" and make it clear for everyone what it tries to model: 
* immune response from the interfacing with flesh/wounds
* inflammation from crossing the skin barrier
* other adverse effects from influence on the body inner workings

Important, it should be clear that "invasiveness" **is not** tied to the volume of the CBM in the general case. The volume/weight should be modeled by encumbrance.

So, CBMs should have two main penalties:
* "invasiveness"  per bodyparts, limited by cap
* encumbrance, per bodypart

Then all CBMs should have their penalties adjusted according to the new definition. 

To address the question raised by Asmitha90:
> If we have bionic like "Repair Nanobots" then i assume we reached nano technology level. This would make most of the surgery, safe space or incompatibility problem non-existent(neural stuff and similar would still be problematic) since the surgeries cloud happen at cell level.

The "invasiveness"  part of the CBMs should be set with having the alien technologies in mind, meaning, it could be lower that it would be IRL today, if it fits the lore.

**Comment by actual-nh** [comment]
> The "invasiveness" part of the CBMs should be set with having the alien technologies in mind, meaning, it could be lower that it would be IRL today, if it fits the lore.

Quite. Another way to put it - the question is not whether we can do it right now; the question is whether there is a reasonable way to accomplish it (given the available - in-game - technology). The first question is of science and technology; the second question is of hard science fiction.

**Comment by Ramza13** [comment]
So I like the idea of changing slots to invasiveness but max invasiveness doesn't really make sense.  Maybe max encumbrance or something instead?  

I feel some sort of upper limit has to exist.  Like sure you could maybe somehow fit two shotguns in your left arm with tradeoffs but obviously not 200.  And especially when replacing bodyparts full scale is possible there has to be some sort of cap.

**Comment by Venera3** [comment]
I'm not that keen on the word "invasiveness" to be honest. "Slots" communicates the main point well, and intuitively implies a maximum. The invasiveness of the cbms that get adjusted higher than what their volume would imply can be mentioned in their description, if it's absolutely necessary.

**Comment by Aivean** [comment]
> I'm not that keen on the word "invasiveness" to be honest. "Slots" communicates the main point well, and intuitively implies a maximum. The invasiveness of the cbms that get adjusted higher than what their volume would imply can be mentioned in their description, if it's absolutely necessary.

I'm not adamant on using the term "invasiveness", but to me the term "slot" conveys a very strong signal that it's a pre-existing **space**, which could be filled with bionic. I'm afraid that the numbers for "slots" for current and new bionics would be derived from bionic volume, which, I believe, is wrong.

Again, the problem that I'm trying to solve here is this:
* volume of the CBMs should be modeled by encumbrance penalty only
* slots/invasiveness should be used to to model all other adverse effects of CBM installation **sans volume**

Maybe there is an alternative way to standardize the process of deriving these two metrics, without renaming.

**Comment by Venera3** [comment]
Is "capacity" used in any other context with bionics? It doesn't imply the same pre-definition as "slots", while keeping it flexible (and sparing us another "-ness" in the process).

**Pull request closed without merging**